### PR TITLE
Add support for Collection POST,PATCH

### PR DESCRIPTION
**Pull request by sirosen**
This may look rather big, as it adds ~500 physical lines to the `src/` tree, but a *lot* of that is either documentation or mindlessly repetitive lists of options. In terms of logical lines, I think there's less than 50 new lines added.

This adds
- `GCSClient.create_collection`
- `GCSClient.update_collection`
- `MappedCollectionDocument`
- `GuestCollectionDocument`

Importantly, the collection document helpers automatically deduce `DATA_TYPE` (more on this below) based on the fields which are present. This makes it ~trivial~ ~easy~ straightforward to write code which is compatible with a broad range of GCS API versions for these operations, even if (in the future) new document versions are added. In particular, I wanted to make sure we're well prepared if we need to do a CLI update in a hurry, without doing a corresponding SDK release.

On that note... these changes are needed for collection create and update operations in the CLI. Without the `DATA_TYPE` deduction work here, the CLI would need a repeat of these same efforts, which is obviously undesirable. I think we should get this in and cut another beta so that we can use this in the CLI `main` right away.

---

These are the create and update operations for GCS Collections.
There is also API support for PUT, but as it is harder to use correctly than PATCH, and does not support operations which cannot be accomplished via the PATCH endpoint, no SDK method is added for it here. One can easily be added later if necessary.

With respect to the payload helpers, this introduces three new classes: a base class, and a class for each of Mapped and Guest collection types.

Because the same payload helpers will be used for POST and PATCH behaviors, none of the arguments are required. Documentation covers the fact that some arguments are needed during creation.

Many arguments are repeated verbatim between the mapped and guest collection types and their base class. For this reason, the `dead` code analyzer is particularly helpful. Unlike code coverage analysis, or `pycqa`'s unused variable checks, `dead` will insist that each variable in a function definition is used. However, `dead` is easily misled, e.g. by signature lines which are needed to match some external specification. For this reason, `dead` is being enabled as a
selective linting step, only against the GCS helper module for now.

Some unusual conventions around commenting on options and specifying an ordering are used to ensure that the module is internally consistent, and they are documented in a block comment in that module.

GCS API requires that the `DATA_TYPE` field be set based on the newest field used in a payload. That is, if `foo` was added in v1.5.0, then the `DATA_TYPE` field must be v1.5.0 or higher. However, SDK users and clients (e.g. globus-cli) will want to maintain compatibility with as many GCS servers as is possible at any given time. For this reason, `DATA_TYPE` should not be aggressively set to the highest known version. In support of these requirements, `CollectionDocument` will inspect the fields which are present and set `DATA_TYPE` to the lowest possible version number. In support of new fields which require higher version numbers, this is exposed as a modifiable mapping on the CollectionDocument, `DATATYPE_VERSION_IMPLICATIONS`. Additionally, `DATA_TYPE` can be passed explicitly to set a known value. This usage is to be considered secondary to use of `DATATYPE_VERSION_IMPLICATIONS`, which is the preferred way to set `DATA_TYPE`.
See `test_datatype_version_deduction` and related tests for a demonstration of this behavior.

### CODE REVIEW

**Review by aaschaer — COMMENTED** [review]
One thing I'm concerned about is how DATATYPE_VERSION_IMPLICATIONS will handle fields that are no longer supported in future versions, since I could imagine a scenario where the sdk needs to support `field_dropped_in_2.0` and `field_added_in_2.1`, but passing both of them would create an invalid document.
Maybe its fine to just set the version to 2.1 in that case and let GCS return an error?
Or maybe I shouldn't even worry about it until its actually an issue

Other than that code looks good to me

File: src/globus_sdk/services/gcs/data.py [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+# locations, so we need to impose some order on things to make comparisons easy.
+#
+# Complicating this, there are some arguments to specific types which aren't shared
+# by the common base.
+#
+# The rule and rationale used is as follows:
+# - DATA_TYPE is special, and always first
+# - next, the common optional arguments (shared by all)
+# - after that, the specific optional arguments for this type/subtype
+# - 'additional_fileds' is special, and always lastA
```
[the remark this answers is not in the document]
Comment (sirosen) [reply src/globus_sdk/services/gcs/data.py:22]: Alack! My `vim` is showing! (`A` is "insert at end of line; clearly double-tapped it here.) Will fix in a moment.

**Comment by sirosen** [comment]
> ... a [future] scenario where the sdk needs to support `field_dropped_in_2.0` and `field_added_in_2.1`, but passing both of them would create an invalid document.
> Maybe its fine to just set the version to 2.1 in that case and let GCS return an error?

Yes, to me this seems fine. The key thing is that we get the following:
```
MappedCollectionDocument(X1...XN) -> {"DATA_TYPE": "collection#1.0.0"}
MappedCollectionDocument(X1...XN, field_added_in_21="foo") -> {"DATA_TYPE": "collection#2.1.0"}
```

This means that code using arguments `X1...XN` will be compatible with any GCS which accepts a v1.0 document. But the moment the `field_added_in_21` arg is added, the minimum required GCS version jumps up to 2.1.0 .

If you pass both, we'd get a version of `2.1.0`, and then `field_dropped_in_2.0` would be rejected by GCS as an unknown field.
The usage is really invalid, so the only alternative would be for `CollectionDocument` to catch this for you immediately.

Setting aside what the implementation looks like, if you write
```python
# shorten the names a bit...
MappedCollectionDocument(drop20="foo", add21="bar")
```

what is the desired behavior / outcome? Some are very easy to rule out as "that makes no sense".

1. `ValueError` (or a subclass thereof)

This is ideal. We know that the combination is invalid. It's possible _at least in theory_ for the SDK to know too.

2. `{"DATA_TYPE": "collection#2.1.0", "drop20": "foo", "add21": "bar"}`

This is the current behavior. The document is invalid, but it won't be known as such until it's submitted to GCS.

---

So why not choose `ValueError` as the behavior if we know that it's better?

I think the implementation quickly gets out of hand. For each new scenario we handle, a new corner case can arise.

For example, imagine
```python
MAX_DATATYPE_VERSION_IMPLICATIONS = {
    "drop20": (2, 0, 0),
}
```

What happens if `"drop20"` is re-added in version 2.5.0? Then we need it to be
```python
INVALID_DATATYPE_VERSION_IMPLICATIONS = {
    "drop20": ((2, 0, 0), (2, 5, 0)),
}
```

And what if `"drop20"` is only valid in versions `3.0.0+` if the field `"add30"` is also present?
```python
INVALID_DATATYPE_VERSION_IMPLICATIONS = {
    "drop20": (((2, 0, 0), (2, 5, 0)), lambda x, v: v < (3,) or "add30" in x),
}
```

In summary: payload validation is an arbitrary task. If you want to go down this road, sooner or later you want to have a spec document or schemas to check against.

I'm 100% in favor of us adding this validation at some point, but it will have to be thought about as a whole new feature.
Ideally, there would be an openapi validator like `jsonschema` which we could hook into things and feed the specs, but `jsonschema` (the library) can't handle the openapi extensions to jsonschema (the spec).

**Comment by sirosen** [comment]
> Other than that code looks good to me

On the basis of this I intend to merge once the build with my fixup passes.

If I'm not convincing on why we shouldn't have upper bounds or other things, let me know.

My main concern is about how we'll handle `globus collection update ...` in the CLI. We'll want that to be "as compatible as possible" at all times.
